proposed/simple-csharp-programs-motivation-walkthrough.md

```diff
@@ -0,0 +1,311 @@
+
+```csharp
+#!/usr/bin/share/dotnet
+#r "framework:aspnet"
+
+var app = WebApplication.CreateBuilder(args);
+app.MapGet("/", () => "Hello World");
+await app.RunAsync();
+```
+
+### Shebangs and their semantics
+
+The shebangs in the walkthrough all point at `/usr/bin/share/dotnet`. Substitute a path for your appropriate installation of .NET on your operating system of choice.
+
+What this means is that it identifies the .NET environment and implies a default set of references available in the file. However, it does not bring in other things like ASP.NET or other frameworkreferences. It is a contextual, "this is how I run" sort of directive.
+
+It could be extended/changed in a few ways:
+
+* `/usr/bin/share/dotnet run` makes it explicit that the command is to run the file
+* `/usr/bin/share/dotnet-aspnet` is an alternative to `#!aspnet` that brings ASP.NET Core into scope for the file (substitute a different framework/sdk here)
+
+In effect, the shebang's semantics could span a spectrum of "I need the path and command to literally execute this file" towards "I add context that affects the references available to this file".
+
+We'll need to decide if they can affect the set of references available in the file or not.
+
+### Build output
+
+The walkthrough states that the build output is a single executable file in the current working directory. This is based on the hypothesis that this is the least surprising place to put the executable.
+
+* Is this hypothesis based on reality?
+* Is this still true if you pull in a package?
+* Is this still true if you pull in a framework? (e.g., ASP.NET Core is hundreds of assemblies, where do they go?)
+
+### Nuget reference syntax and naming
+
+The current walkthrough proposed `"#r "nuget:..."` for packages.
+
+Syntax could change:
+
+* Use a `#!r` shebang-style syntax
+* Use a custom directive like `#nuget`
+
+Naming could change:
+
+* `package` instead of `nuget`
+
+However, that seems unlikely given that the current syntax is already shipping in .NET Interactive.
+
+### A note on .NET Interactive
+
+.NET Interactive _may_ be able to service these goals, but since it is currently focused on notebooks-based programming it doesn't really address the same scenario like JavaScript, Python, and Ruby. It is highly relevant for people who are used to notebook environments, and since notebooks are breaking into educational spaces, it may also play a big role in getting people started there.
+
+But if a JS/Python/Ruby programmer is looking to try out C# for the first time, the way that they know how to start with a language is not well supported by our tooling.
+
+### A note on F#
+
+With some very real ["Simpsons already did it"](https://knowyourmeme.com/memes/the-simpsons-did-it) energy, we here recognize that F# has already gone down this path with F# scripting and F# Interactive:
+
+**hello.fsx**:
+
+```fsharp
+printfn "Hello, World!"
+```
+
+and run it:
+
+```bash
+$ dotnet fsi hello.fsx
```

Comment:
If we were to re-bolt the design proposed in this document onto F#, what would the shebang line look like?

Reply:
A "Simple F# Programs" proposal would probably work similarly, assuming that F# supports shebang syntax and global `using`s inferred by the build environment. I don't think there's anything stopping F# from using the same syntax.